File: UnityMcpServer/src/tools/compile_and_reload.py
from mcp.server.fastmcp import FastMCP, Context
from typing import Dict, Any
from unity_connection import get_unity_connection

def register_compile_and_reload_tools(mcp: FastMCP):
    """Register compile and reload tools with the MCP server."""

    @mcp.tool()
    def compile_and_reload(
        ctx: Context,
        action: str = "compile_and_reload",
    ) -> Dict[str, Any]:
        """Manually trigger Unity script compilation and domain reload.

        Args:
            action: Operation to perform:
                - 'compile': Only compile scripts
                - 'reload': Only reload domain
                - 'refresh': Only refresh assets
                - 'compile_and_reload': Full compile and reload (default)

        Returns:
            Dictionary with operation results.
        """
        try:
            params = {
                "action": action,
            }
            
            response = get_unity_connection().send_command("compile_and_reload", params)
            
            if response.get("success"):
                return {
                    "success": True,
                    "message": response.get("message"),
                    "data": response.get("data")
                }
            else:
                return {
                    "success": False,
                    "message": response.get("error", "Unknown error occurred")
                }
                
        except Exception as e:
            return {
                "success": False,
                "message": f"Python error: {str(e)}"
            }

File: UnityMcpBridge/Assets/Scripts/Editor/Tools/CompileAndReload.cs
using System;
using UnityEditor;
using UnityEditor.Compilation;
using UnityEngine;
using Newtonsoft.Json.Linq;
using UnityMcpBridge.Editor.Helpers;

namespace UnityMcpBridge.Editor.Tools
{
    public static class CompileAndReload
    {
        public static object HandleCommand(JObject @params)
        {
            string action = @params["action"]?.ToString().ToLower();
            if (string.IsNullOrEmpty(action))
            {
                return Response.Error("Action parameter is required.");
            }

            try
            {
                switch (action)
                {
                    case "compile":
                        return CompileScripts();
                    case "reload":
                        return ReloadDomain();
                    case "refresh":
                        return RefreshAssets();
                    case "compile_and_reload":
                        return CompileAndReloadAll();
                    default:
                        return Response.Error($"Unknown action: '{action}'. Valid actions: compile, reload, refresh, compile_and_reload");
                }
            }
            catch (Exception e)
            {
                Debug.LogError($"[CompileAndReload] Action '{action}' failed: {e}");
                return Response.Error($"Internal error: {e.Message}");
            }
        }

        private static object CompileScripts()
        {
            try
            {
                Debug.Log("[CompileAndReload] Requesting script compilation...");
                CompilationPipeline.RequestScriptCompilation();
                return Response.Success("Script compilation requested successfully.");
            }
            catch (Exception e)
            {
                return Response.Error($"Failed to compile scripts: {e.Message}");
            }
        }

        private static object ReloadDomain()
        {
            try
            {
                Debug.Log("[CompileAndReload] Requesting domain reload...");
                EditorUtility.RequestScriptReload();
                return Response.Success("Domain reload requested successfully.");
            }
            catch (Exception e)
            {
                return Response.Error($"Failed to reload domain: {e.Message}");
            }
        }

        private static object RefreshAssets()
        {
            try
            {
                Debug.Log("[CompileAndReload] Refreshing assets...");
                AssetDatabase.Refresh();
                return Response.Success("Assets refreshed successfully.");
            }
            catch (Exception e)
            {
                return Response.Error($"Failed to refresh assets: {e.Message}");
            }
        }

        private static object CompileAndReloadAll()
        {
            try
            {
                Debug.Log("[CompileAndReload] Performing full compile and reload...");
                
                // First refresh assets
                AssetDatabase.Refresh();
                
                // Then request compilation
                CompilationPipeline.RequestScriptCompilation();
                
                // Finally request domain reload
                EditorUtility.RequestScriptReload();
                
                return Response.Success("Full compile and reload requested successfully.");
            }
            catch (Exception e)
            {
                return Response.Error($"Failed to compile and reload: {e.Message}");
            }
        }
    }
}